### PR TITLE
Introduce Requirements struct to clarify code

### PR DESCRIPTION
**Pull request by djc**
`cargo::core::resolver::build_requirements()` previously, in this order:

* Defined local variables to track state
* Called a function to mutate the local variables
* Defined the aforementioned function
* Returned two out of three local variables as a tuple

This PR changes the code so that this is a recast as a struct (appropriately called `Requirements`), which is mutated in a more fine-grained way by its methods and acts as the return type for `build_requirements()`. To me, this seems a lot easier to understand.

This work was done in the context of #1286, but I figured it was easier to start landing some of the refactoring to avoid bitrot and get early (well, earlier) feedback.

### CODE REVIEW

**Comment by djc** [comment]
@alexcrichton this is what I started with at RustFest. I have some more work done on using features as a type, but that requires yet more untangling of all the related code passing around features as strings.

File: src/cargo/core/resolver/mod.rs [hunk 1]
```diff
@@ -859,55 +859,68 @@ fn compatible(a: &semver::Version, b: &semver::Version) -> bool {
...
-    match *method {
-        Method::Everything |
-        Method::Required { uses_default_features: true, .. } => {
-            if s.features().get("default").is_some() {
-                add_feature(s, "default", &mut deps, &mut used,
-                            &mut visited)?;
+
+    fn require_dependency(&mut self, pkg: &'r str) {
+        if self.seen(pkg) {
+            return;
```
Comment (alexcrichton) [inline src/cargo/core/resolver/mod.rs]: I think this early return may cause the boolean in the `deps` entry to be `false` when it should unconditionally be `true` if this function is called, right?

File: src/cargo/core/resolver/mod.rs [hunk 2]
```diff
@@ -859,55 +859,68 @@ fn compatible(a: &semver::Version, b: &semver::Version) -> bool {
...
+            visited: HashSet::new(),
         }
-        Method::Required { features: requested_features, .. } =>  {
-            for feat in requested_features.iter() {
-                add_feature(s, feat, &mut deps, &mut used, &mut visited)?;
-            }
+    }
+
+    fn require_crate_feature(&mut self, package: &'r str, feat: &'r str) {
+        self.used.insert(package);
```
Comment (alexcrichton) [inline src/cargo/core/resolver/mod.rs]: I'd naively expect this to call `require_dependency` and then fill in the `deps` map list, but that also seems like the kind of thing that may break tests...

File: src/cargo/core/resolver/mod.rs (hunk 1 above)
Comment (djc) [reply src/cargo/core/resolver/mod.rs]: That's also what happens in the old version, right? I tried to keep it functionally the same; reviewing it now, this seems to be the behavior of the old code, too.

File: src/cargo/core/resolver/mod.rs (hunk 2 above)
Comment (djc) [reply src/cargo/core/resolver/mod.rs]: I'm still a little fuzzy on what `used` means exactly, and how it's different from the `deps` boolean being `true`. This is different at least in that it leaves the `deps` boolean at `false`, which seems meaningful because setting a "forwarded" feature does not mean that we should actually pull that dependency in.

File: src/cargo/core/resolver/mod.rs (hunk 1 above)
Comment (djc) [reply src/cargo/core/resolver/mod.rs]: (Also, wouldn't you say that the boolean in `deps` must already be true if `pkg` is `self.seen()` here?)

File: src/cargo/core/resolver/mod.rs (hunk 2 above)
Comment (alexcrichton) [reply src/cargo/core/resolver/mod.rs]: I think it may be related to https://github.com/rust-lang/cargo/pull/4364? I think that changed a few things in this area

File: src/cargo/core/resolver/mod.rs (hunk 1 above)
Comment (alexcrichton) [reply src/cargo/core/resolver/mod.rs]: Ah ok I think I see that now, makes sense! I find this code always pretty hard to follow...

I completely forget at this point why these all exist as they are, this has changed a fair bit since first written.
